File: Server/Repository/src/main/java/org/elkoserver/server/repository/Repository.java
package org.elkoserver.server.repository;

import org.elkoserver.foundation.actor.RefTable;
import org.elkoserver.foundation.json.AlwaysBaseTypeResolver;
import org.elkoserver.foundation.server.Server;
import org.elkoserver.objdb.ObjectStoreFactory;
import org.elkoserver.objdb.store.ObjectStore;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

/**
 * Main state data structure in a Repository.
 */
class Repository {
    /** Table for mapping object references in messages. */
    private RefTable myRefTable;

    /** Server object. */
    private Server myServer;

    /** Local object storage module. */
    private ObjectStore myObjectStore;

    /** Number of repository clients currently connected. */
    private int myRepClientCount;

    /** Flag that is set once server shutdown begins. */
    private boolean amShuttingDown;

    /**
     * Constructor.
     *
     * @param server  Server object.
     * @param appTrace  Trace object for diagnostics.
     */
    Repository(Server server, Trace appTrace, TraceFactory traceFactory) {
        myServer = server;

        myObjectStore = ObjectStoreFactory.createAndInitializeObjectStore(server.props(), "conf.rep", appTrace);

        myRefTable = new RefTable(AlwaysBaseTypeResolver.theAlwaysBaseTypeResolver, traceFactory);
        myRefTable.addRef(new RepHandler(this, traceFactory));
        myRefTable.addRef(new AdminHandler(this, traceFactory));

        amShuttingDown = false;
        myRepClientCount = 0;
        server.registerShutdownWatcher(() -> {
            amShuttingDown = true;
            countRepClients(0);
        });
    }

    /**
     * Add to or subtract from the number of repository clients connected.  If
     * the number drops to zero and this server is in the midst of shutting
     * down, terminate the object store.
     *
     * @param delta  The amount to change the count by.
     */
    void countRepClients(int delta) {
        myRepClientCount += delta;
        if (amShuttingDown && myRepClientCount <= 0) {
            myObjectStore.shutdown();
        }
    }

    /**
     * Test if the server is in the midst of shutdown.
     *
     * @return true if the server is trying to shutdown.
     */
    boolean isShuttingDown() {
        return amShuttingDown;
    }

    /**
     * Get the object store currently in use.
     *
     * @return the object storage object for this server.
     */
    ObjectStore objectStore() {
        return myObjectStore;
    }

    /**
     * Get the ref table.
     *
     * @return the object ref table that resolves object reference strings for
     *    messages sent to this server.
     */
    RefTable refTable() {
        return myRefTable;
    }

    /**
     * Reinitialize the server.
     */
    void reinit() {
        myServer.reinit();
    }

    /**
     * Shutdown the server.
     *
     * @param kill  If true, shutdown immediately instead of cleaning up.
     */
    void shutdown(boolean kill) {
        myServer.shutdown(kill);
    }
}

File: Server/Repository/src/main/kotlin/org/elkoserver/server/repository/Repository.kt
package org.elkoserver.server.repository

import org.elkoserver.foundation.actor.RefTable
import org.elkoserver.foundation.server.Server
import org.elkoserver.foundation.server.ShutdownWatcher
import org.elkoserver.objectdatabase.store.ObjectStore
import org.elkoserver.util.trace.slf4j.Gorgel

/**
 * Main state data structure in a Repository.
 *
 * @param myServer Server object.
 */
internal class Repository(
    private val myServer: Server,
    internal val myRefTable: RefTable,
    baseCommGorgel: Gorgel,
    internal val myObjectStore: ObjectStore,
    shutdownWatcher: ShutdownWatcher
) {

    /** Number of repository clients currently connected.  */
    private var myRepClientCount = 0

    /**
     * Test if the server is in the midst of shutdown.
     *
     * @return true if the server is trying to shutdown.
     */
    /** Flag that is set once server shutdown begins.  */
    var isShuttingDown = false

    /**
     * Add to or subtract from the number of repository clients connected.  If
     * the number drops to zero and this server is in the midst of shutting
     * down, terminate the object store.
     *
     * @param delta  The amount to change the count by.
     */
    fun countRepClients(delta: Int) {
        myRepClientCount += delta
        if (isShuttingDown && myRepClientCount <= 0) {
            myObjectStore.shutDown()
        }
    }

    /**
     * Reinitialize the server.
     */
    fun reinit() {
        myServer.reinit()
    }

    fun shutDown() {
        isShuttingDown = true
        countRepClients(0)
    }

    init {
        myRefTable.addRef(RepHandler(this, baseCommGorgel.getChild(RepHandler::class)))
        myRefTable.addRef(AdminHandler(this, shutdownWatcher, baseCommGorgel.getChild(AdminHandler::class)))
    }
}
